Support bisect in spec opts

urbanautomaton opened the pull request:
I'm breaking out a new PR to fix #2231, based on discussion in #2259.

To summarise progress: the infinite looping problem is solved by modifying the environment given to the child processes, so that only the parent process starts a bisection.

However, the existing invocation of `Bisect::Runner` from within `OptionParser` is problematic, because `OptionParser` is invoked twice: once for CLI args, and once for args set in the environment. We need to invoke the bisection only once *all* args have been parsed, so that the correct commands can be issued to the child processes.

@myronmarston has suggested we refactor the `OptionParser` so it no longer invokes any actions directly, but instead returns a callable that the main runner can invoke. The current diff contains a sketch implementation of what I understand this to mean.

@myronmarston - is this sort of thing what you envisaged? (Obviously there'd be further refactoring to do, I just wanted to check I've got the right general idea...)

myronmarston commented:
Hey @urbanautomaton, thanks for taking a stab at this.  It definitely looks like a good start.

I think the refactoring of option parsing can be a self-contained PR.  What do you think about opening a PR for just that so we can get it merged quick?  Then you can deal with the bisect issues in a PR on top of that.

I'll go leave a few comments about the option parsing stuff on this PR now as well.

myronmarston commented on lib/rspec/core/configuration_options.rb:
```diff
@@ -13,6 +13,10 @@ def initialize(args)
         organize_options
       end
 
+      def command_line_and_env_args
```
It's not clear to me why we need a method containing both command line and env args but not args from the dot files.  Can you explain?

urbanautomaton replied:
It's because I was transferring all the env args to the CLI args of the child processes and removing SPEC_OPTS completely for the children, but there's actually no need to do that, so this can be removed.

myronmarston commented on lib/rspec/core/option_parser.rb:
```diff
@@ -68,7 +68,17 @@ def parser(options)
         parser.on('--bisect[=verbose]', 'Repeatedly runs the suite in order to isolate the failures to the ',
                   '  smallest reproducible case.') do |argument|
-          bisect_and_exit(argument)
+          options[:runner] = -> (opts, err, out) do
```
We still need to support 1.8.7 so you'll have to use `Proc.new` or `lambda` instead of the 1.9+ stabby lambda syntax.

myronmarston commented on lib/rspec/core/option_parser.rb:
```diff
@@ -97,15 +107,27 @@ def parser(options)
-          initialize_project_and_exit
+          options[:runner] = -> (*_args) do
+            initialize_project_and_exit
+          end
```
Could this be `options[:runner] = method(:initialize_project_and_exit)`? The method would have to take the args, but I think that's ok.

myronmarston commented on lib/rspec/core/option_parser.rb:
```diff
@@ -68,7 +68,17 @@ def parser(options)
+            )
+
+            exit(success ? 0 : 1)
+          end
```
I think it's a bit easier to read if all the runner actions have a method that implements it, so that the parsing logic here can just be `options[:runner] = method(:bisect_and_exit)`.

JonRowe commented on features/command_line/bisect.feature:
```diff
@@ -155,3 +155,30 @@ Feature: Bisect
+      """
+    When I run `rspec ./spec/calculator_10_spec.rb[1:1] ./spec/calculator_1_spec.rb[1:1] --seed 1234`
+    Then the output should contain "2 examples, 1 failure"
+
```
Is this temporary?

urbanautomaton replied:
I think so, yes - I lobbed it in to get something to work against, but without my fix this test will forkbomb any machine it runs on, so I don't think it should stick around. 😄

Now that it's working, I can update the assertions made about the `Bisect::Runner` commands to check that `--bisect` is dropped from the env passed to `Open3` - I think that should be sufficient.

myronmarston replied:
Agreed we should not keep a cuke as its documentation value is very low.  This could be converted to an integration spec (in [this file](https://github.com/rspec/rspec-core/blob/master/spec/integration/bisect_spec.rb)) if you want to keep integration coverage of this. But hopefully the unit specs about how `ENV` is handled is sufficient.

urbanautomaton commented:
@myronmarston Thanks for the feedback. Responding inline here so the convo doesn't get lost when the diff changes:

>     options[:runner] = -> (opts, err, out) do
>       initialize_project_and_exit
>     end
>
> Could this be options[:runner] = method(:initialize_project_and_exit)? The method would have to take the args, but I think that's ok.

I had a stab at this, but ran into difficulties when porting `#print_help_and_exit`, because it requires extra information that's local to the parser, and not accessible via the options passed to the callable.

I've now tried something similar to what you suggest, but with factory methods returning lambdas that can capture any information necessary for the particular callable. What do you think?

myronmarston commented on lib/rspec/core/bisect/runner.rb:
```diff
@@ -89,6 +89,10 @@ def run_command(cmd)
         end
 
+        def spec_opts_without_bisect
+          { 'SPEC_OPTS' => ENV.fetch('SPEC_OPTS', '').gsub(/--bisect/, '') }
```
What if the user has put `--bisect=verbose` in `SPEC_OPTS`?  This gsub will leave `=verbose`.

myronmarston commented on lib/rspec/core/option_parser.rb:
```diff
@@ -282,38 +283,64 @@ def configure_only_failures(options)
-      exit
-    end
+    module Callables
+      class << self
```
I think I'd prefer if this just used `def self.` for each method or used `module_function`.

myronmarston commented:
> I've now tried something similar to what you suggest, but with factory methods returning lambdas that can capture any information necessary for the particular callable. What do you think?

:+1:

urbanautomaton commented:
I've addressed the feedback so far, and replaced the cuke with assertions at the `Bisect::Runner` unit test level. I think I'm happy with that, and don't feel like integration-level testing is necessary.

For JRuby and 1.8.7, `Open3.popen3` doesn't support a passed environment. I've fallen back to modifying-and-restoring the env of the parent process in this case; I felt this was less error-prone than trying to build a command string with environment variables in it. I'd appreciate feedback on whether the assertions this necessitates are too complex as a result.

At the moment I'm testing the runner callables as part of the `OptionParser` tests, and have consolidated these tests there ([moving some tests](https://github.com/rspec/rspec-core/pull/2271/commits/965e2d16553899a1dd002a053418808ca1e45096) of the `DRbRunner` invocation as appropriate). Does this seem reasonable? I'm half-tempted to extract the callables to their own file and test them directly - would renaming the module something like `RSpec::Core::Invocations` make sense? (I would suggest `Runners`, but this name is pretty taken.)

The suite is almost totally green, but for a single warning generated by [the 2.2 job](https://travis-ci.org/rspec/rspec-core/jobs/137337547). This job is running 2.2.0-p0, and the warning seems to be in error, as it doesn't appear in later versions of the 2.2.x branch. Is testing against 2.2.0-p0 deliberate, or could this job be updated to 2.2.5? If the former, I'd love a pointer as to how to address this warning, because I can't see anything wrong with the code that's triggering it.

myronmarston commented:
> The suite is almost totally green, but for a single warning generated by the 2.2 job. This job is running 2.2.0-p0, and the warning seems to be in error, as it doesn't appear in later versions of the 2.2.x branch. Is testing against 2.2.0-p0 deliberate, or could this job be updated to 2.2.5? If the former, I'd love a pointer as to how to address this warning, because I can't see anything wrong with the code that's triggering it.

This was a warning added to 2.2.0:

https://bugs.ruby-lang.org/issues/10661

Ruby-core removed the warning when I raised that issue, and it got removed from 2.2.1.  (Unfortunately, it was added at the last minute to 2.2.0 and so it was never part of any of the RCs so I could notice and comment then).

To fix the warning on 2.2.0, you'll need to apply one of these fixes:

* Rename the `args` local variable from [earlier in the file](https://github.com/urbanautomaton/rspec-core/blob/21ed282ce87103c9b5db4fd06056d13ffde997c8/spec/rspec/core/option_parser_spec.rb#L32)
* Rename the [`let(:args)`](https://github.com/urbanautomaton/rspec-core/blob/21ed282ce87103c9b5db4fd06056d13ffde997c8/spec/rspec/core/option_parser_spec.rb#L417)
* Change the calls to the `args` method (defined by the `let`) so that they are obviously method calls and not variables -- e.g. `args()` or `self.args`.

All the fixes for this are sub-optimal and the warning is annoying (it's why I raised the issue on ruby-core when I first noticed).

> Is testing against 2.2.0-p0 deliberate, or could this job be updated to 2.2.5?

Testing against 2.2.0-p0 is _not_ deliberate. I thought that are use of `2.2` in `.travis.yml` would cause travis to pick the _latest_ available version of 2.2 (something like 2.2.5) but that does not work.  So I think it's worth changing it to `2.2.5` in `.travis.yml` so we don't have to deal with the annoying warning anymore (it has come up multiple times, unfortunately).

myronmarston commented on spec/rspec/core/bisect/runner_spec.rb:
```diff
@@ -213,6 +213,27 @@ def repro_command_from(ids)
+            expect(Open3).to have_received(open3_method).with(command)
+            expect(called_environment).to include(environment)
+          end
+        end
```
If `open3_method` is set to something besides `capture2e`, the `match` logic does nothing, which could lead to a silent failure.  I think I'd prefer to see this as:

``` ruby
if open3_method == :capture2e
  RSpec::Matchers.define :invoke_command_with_env do |command, environment|
    # ...
  end
elsif open3_method == :popen3
  RSpec::Matchers.define :invoke_command_with_env do |command, environment|
    # ...
  end
end
```

Then, if a different method is used, the matcher won't even be defined and we'll definitely get an error in that case.

myronmarston commented on spec/rspec/core/bisect/runner_spec.rb:
```diff
@@ -213,6 +213,27 @@ def repro_command_from(ids)
+
+        def supports_block_expectations?
+          true
+        end
```
This can just be `supports_block_expectations` since you are in a `Matchers.define` block.

myronmarston commented on spec/rspec/core/bisect/runner_spec.rb:
```diff
@@ -213,6 +213,27 @@ def repro_command_from(ids)
+            called_environment = nil
+            allow(Open3).to receive(open3_method) do
+              called_environment = ENV.to_hash.dup
+            end
```
It's a little weird that you stub this here but not for the `capture2e` branch.  In the `before` hook below I see that this method is also stubbed, but then I get to wondering how these two stubs interact. Is there a way we can clean it up and make it more consistent?

myronmarston replied:
For example, maybe we can update the stub in the `before` hook below so has this stub implementation (besides the canned return value) and always captures the called environment?

Then this matcher would not even need to use the block form, right?

urbanautomaton replied:
Ah, good point - I've cleaned up the stubbing. I'm probably being daft here but I can't figure out what the assertion would look like in non-block form, though (I'm unsure what the "actual" value would be to read well):

```ruby
it "runs the command" do
  runner.original_results

  expect(???).to have_invoked_command_with_env(a_string_including("--seed 1234"), {})
end
```

`expect(a_string_including("--seed 1234").to have_been_invoked_with_env({})` perhaps?

myronmarston commented on spec/rspec/core/option_parser_spec.rb:
```diff
@@ -81,28 +83,101 @@ def generate_help_text
+    %w[ -X --drb ].each do |option|
+      describe option do
+        let(:parser) { Parser.new([option]) }
+        let(:options) { double(:options) }
```
It's not clear to me what this `let` is used for.  All the examples below I've looked at so far have a line like `options = ...` -- meaning they shadow this with a local variable and don't even use it.  Is this needed, or can it be moved into just the examples that need it?

myronmarston commented on spec/rspec/core/option_parser_spec.rb:
```diff
@@ -81,28 +83,101 @@ def generate_help_text
+        let(:parser) { Parser.new([option]) }
+        let(:options) { double(:options) }
+        let(:err) { double(:stderr, :puts => nil) }
+        let(:out) { double(:stderr, :puts => nil) }
```
I'd prefer to see verifying doubles used here:

``` ruby
let(:err) { object_double($stderr, :puts => nil) }
let(:out) { object_double($stdout, :puts => nil) }
```

myronmarston replied:
Actually, the interface for IO objects is quite large (making doubles that stand-in for an IO pretty brittle) and ruby provides a perfectly usable double that _exactly_ mimics the full interface in `StringIO` -- so I think I'd prefer to see a `StringIO` used here instead.

myronmarston commented on spec/rspec/core/option_parser_spec.rb:
```diff
@@ -81,28 +83,101 @@ def generate_help_text
+
+        context 'when a DRb server is running' do
+          it "builds a DRbRunner and runs the specs" do
+            drb_proxy = double(RSpec::Core::DRbRunner, :run => 0)
```
This should ideally use `instance_double` as well -- we prefer to use them whenever possible since they provide stronger guarantees.

myronmarston commented:
:+1: Looking good, @urbanautomaton!  We are expecting to release 3.5 in the next few days (we plan to release as soon as rails 5 final ships, and we hear it should be this week), so I'm hoping we can get this merged before 3.5 ships :).

myronmarston commented:
> At the moment I'm testing the runner callables as part of the OptionParser tests, and have consolidated these tests there (moving some tests of the DRbRunner invocation as appropriate). Does this seem reasonable? I'm half-tempted to extract the callables to their own file and test them directly - would renaming the module something like RSpec::Core::Invocations make sense? (I would suggest Runners, but this name is pretty taken.)

What you've done is fine.  Your idea to take it a step further with `RSpec::Core::Invocations` sounds like a nice further improvement.  I don't consider it a pre-requisite to merge this PR but if you want to do that, please take a stab at it!

bf4 commented on lib/rspec/core/bisect/runner.rb:
```diff
@@ -71,24 +72,45 @@ def run_locations(*capture_args)
 
+            original_spec_opts = ENV['SPEC_OPTS']
+            ENV['SPEC_OPTS'] = spec_opts_without_bisect
+
```
👍 did you try using the open3 feature of [passing in the env opts](http://ruby-doc.org/stdlib-2.2.3/libdoc/open3/rdoc/Open3.html#method-c-popen3)?

> The parameters env, cmd, and opts are passed to [Process.spawn](http://ruby-doc.org/core-2.2.3/Process.html#method-c-spawn)

> > env: hash
  name => val : set the environment variable
  name => nil : unset the environment variable
> > options: hash
  clearing environment variables:
    :unsetenv_others => true   : clear environment variables except specified by env
    :unsetenv_others => false  : don't clear (default)
> > If a hash is given as env, the environment is updated by env before exec(2) in the child process. If a pair in env has nil as the value, the variable is deleted.

> > set FOO as BAR and unset BAZ.
pid = spawn({"FOO"=>"BAR", "BAZ"=>nil}, command)

> > The :unsetenv_others key in options specifies to clear environment variables, other than specified by env.

> > pid = spawn(command, :unsetenv_others=>true) # no environment variable
pid = spawn({"FOO"=>"BAR"}, command, :unsetenv_others=>true) # FOO only

urbanautomaton replied:
The upside of supporting ruby 1.8.7 is that I get to use the old hash syntax without feeling guilt. The downside is that some of stdlib wasn't as capable back then, and `Open3`/`Process.spawn` is [one of those bits](http://ruby-doc.org/stdlib-1.8.7/libdoc/open3/rdoc/Open3.html), unfortunately. I've used the new passed environment [where possible](https://github.com/urbanautomaton/rspec-core/blob/21ed282ce87103c9b5db4fd06056d13ffde997c8/lib/rspec/core/bisect/runner.rb#L73-L76), but for 1.8.7 it's either put the new env in the command string or mutate `ENV`, so I plumped for the latter.

bf4 commented on spec/rspec/core/bisect/runner_spec.rb:
```diff
@@ -224,8 +245,35 @@ def repro_command_from(ids)
+      end
+
+      context 'when --bisect is present in SPEC_OPTS' do
+        it "runs the suite with --bisect removed from the environment" do
```
you can test the effect of removing `--bisect` from the SPEC_OPTS separately from modifying the SPEC_OPTS when invoking the command , if you wish, using something like:

```ruby
with_env_var 'SPEC_OPTS' => '--bisect --fail-fast' do
  expect(@runner.spec_opts_without_bisect).to eq('--fail-fast')
end
# and then
with_env_vars 'RSPEC_TEMP' => 'original' do
  actual = @runner.run_command "ruby -e 'puts ENV['RSPEC_TEMP']"
  expect(actual).to eq('original')
end
```

urbanautomaton commented:
> Your idea to take it a step further with RSpec::Core::Invocations sounds like a nice further improvement. I don't consider it a pre-requisite to merge this PR but if you want to do that, please take a stab at it!

Okay - given the imminent release, then, I'd like to make that change but will tackle it in a separate PR.

I *think* the build should be passing again now (it's being very slow kicking off for some reason). I'm still not entirely happy with the `Open3` matchers (the way I've got it accessing the `called_environment` seems wrong somehow), but apart from that I think I'm about done. If there's no further feedback I'll rebase the commits into a better story and then we'll hopefully be ready to go.

myronmarston commented on spec/rspec/core/bisect/runner_spec.rb:
```diff
@@ -213,19 +213,67 @@ def repro_command_from(ids)
+            a_string_including("--seed 1234"),
+            { 'SPEC_OPTS' => '--fail-fast' }
+          )
+        end
```
I see what you mean about it being tricky to figure out what to treat as the subject for these expectations.  Maybe the block approach does make the most sense.  I'm happy to merge with this has it is but if you want to switch back to the block form that would be fine, too.

myronmarston commented:
> I think the build should be passing again now (it's being very slow kicking off for some reason).

I think Travis is starting it now.  Must have just gotten backed up for some reason.

> If there's no further feedback I'll rebase the commits into a better story and then we'll hopefully be ready to go.

I left the one comment above about switching back to a block perhaps making more sense and being OK, but whichever approach you decide to take there is fine.  Regardless, this LGTM.  I'll be happy to merge after you've rebased into a clearer story as you said :).

myronmarston commented:
The build is green now, FWIW :).

urbanautomaton commented:
Okay - I went back to the block form for the `Open3` assertions and have rebased the branch into something more coherent.

The `Invocations` extraction turned out pretty easy, so I've done that too (but will happily revert and separately PR it if that commit contains anything controversial at all). Last question: I've marked that whole module `@private` for documentation purposes - is that okay? I wondered whether I should be documenting the module and marking the invocation methods `@private` instead. I don't think it's something any user is likely to access directly, though.

myronmarston commented on spec/rspec/core/bisect/runner_spec.rb:
```diff
@@ -213,19 +213,75 @@ def repro_command_from(ids)
+
+          supports_block_expectations
+        end
+      end
```
Not a merge blocker, but I think maybe the matcher should be named `invoke_command_with_env` instead of the past-tense `have_invoked_command_with_env`.  Consider that all our other block matchers are named with the present tense form (e.g. `change` over `have_changed`, `raise_error` over `have_raised_error`, etc).

urbanautomaton replied:
Oh - yeah, that's an artifact of switching to the non-block style and back again, will change it now.

myronmarston commented on spec/rspec/core/option_parser_spec.rb:
```diff
@@ -317,6 +340,24 @@ def generate_help_text
+
+      it "sets the `:runner` option with the `bisect` invocation" do
+        bisect = double(:bisect)
+        allow(RSpec::Core::Invocations).to receive(:bisect).and_return(bisect)
```
Stubbing invocations to return a double just so we can assert on equality below seems pretty sub-optimal.  As I see it, the `RSpec::Core::Invocations` methods are essentially just exposing constants -- given that they take no arguments.  So WDYT about turning them into constants, and then you don't have to stub here -- instead, down below it can do:

``` ruby
expect(options[:runner]).to eq(RSpec::Core::Invocations::BISECT)
```

Thoughts?

urbanautomaton replied:
The downside there is it doesn't let you capture extra info in the callable ([as required by `.print_help`](https://github.com/urbanautomaton/rspec-core/blob/support-bisect-in-spec-opts/lib/rspec/core/invocations.rb#L46-L52)). I do agree though, I don't particularly like the stubbing. Maybe we could turn them into callable objects?

```ruby
module Invocations
  PrintHelp = Struct.new(:parser, :invalid_options) do
    def call(options, err, out)
       #...
    end
  end
end
```

which would let us assert:

```ruby
expect(options[:runner]).to be_instance_of(RSpec::Core::Invocations::PrintHelp)
```

myronmarston replied:
> The downside there is it doesn't let you capture extra info in the callable (as required by .print_help).

I missed that little detail.  Thanks for pointing it out!

> Maybe we could turn them into callable objects?

:+1:

I like this approach a lot.

urbanautomaton replied:
Done!

myronmarston commented:
:+1: LGTM.  I plan to merge when the build goes green.

urbanautomaton commented:
Missed out the `@private` declarations on the callable classes - patched and now it's green! 🎉

myronmarston commented:
Thanks, @urbanautomaton!

myronmarston commented:
As always, thanks for the great work, @urbanautomaton :).

urbanautomaton commented:
Thanks, and a pleasure as always. 😄
